applications/Preprocessor/lib/gmsh.cpp

```diff
@@ -0,0 +1,446 @@
+    nodes_per_elementtype_[27] =
+        5;  // 5-node fourth order edge (2 nodes associated with the vertices, 3
+    // internal to the edge)
+    dimension_of_elementtype_[27] = 1;
+    nodes_per_elementtype_[28] =
+        6;  // 6-node fifth order edge (2 nodes associated with the vertices, 4
+            // internal to the edge)
+    dimension_of_elementtype_[28] = 1;
+    nodes_per_elementtype_[29] =
+        20;  // 20-node third order tetrahedron (4 nodes associated with the
+             // vertices, 12 with the edges, 4 with the faces)
+    dimension_of_elementtype_[29] = 3;
+    nodes_per_elementtype_[30] =
+        35;  // 35-node fourth order tetrahedron (4 nodes associated with the
+             // vertices, 18 with the edges, 12 with the faces, 1 in the volume)
+    dimension_of_elementtype_[30] = 3;
+    nodes_per_elementtype_[31] =
+        56;  // 56-node fifth order tetrahedron (4 nodes associated with the
+             // vertices, 24 with the edges, 24 with the faces, 4 in the volume)
+    dimension_of_elementtype_[31] = 3;
+    nodes_per_elementtype_[92] =
+        64;  // 64-node third order hexahedron (8 nodes associated with the
+             // vertices, 24 with the edges, 24 with the faces, 8 in the volume)
+    dimension_of_elementtype_[92] = 3;
+    nodes_per_elementtype_[93] =
+        125;  // 125-node fourth order hexahedron (8 nodes associated with the
+              // vertices, 36 with the edges, 54 with the faces, 27 in the
+              // volume)
+    dimension_of_elementtype_[93] = 3;
+}
+
+void GmshReader::readHeader() {
+
+    bool found_header = locate_in_file(Filehandle_, "$MeshFormat");
+    logger.assert_always(found_header, "Header of gmsh file not found");
+    double version;
+    int filetype, datasize;
+    Filehandle_ >> version >> filetype >> datasize;
+
+    logger(VERBOSE, "msh version is % and storage format is %", version,
+           (filetype == 1 ? "binary" : "ASCII"));
+
+    logger.assert_always(version == 2.2,
+                         "Error: Cannot read file versions other than 2.2");
+    logger.assert_always(filetype == 0,
+                         "Error: Cannot read binary format files");
+}
+
+void GmshReader::readNodes() {
+    bool found_nodes = locate_in_file(Filehandle_, "$Nodes");
+    logger.assert_always(found_nodes, "Nodes section of gmsh file not found");
+    size_t number_nodes;
+    Filehandle_ >> number_nodes;
+    size_t nodeTag;
+    double x;
+    double y;
+    double z;
+    nodes_.reserve(number_nodes);
+    for (size_t index = 0; index < number_nodes; index++) {
+        Filehandle_ >> nodeTag >> x >> y >> z;
+        MeshSource2::Coord temp;
+        temp.nodeId =
+            nodeTag - 1;  // gmsh ids are 1 indexed hpgem is zero indexed
+        temp.coordinate = std::vector<double>{x, y, z};
+        nodes_.push_back(temp);
+    }
+}
+
+void GmshReader::readElements() {
+
+    ElementReorder reorder;
+    reorder.addElementType(1, "line:", {0, 1});
```

Comment:
Would again be really nice to have a link here, so that someone could later extend it easily.